src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/TextEditingTargetVisualMode.java

```diff
@@ -0,0 +1,534 @@
+                  if (!loadingFromSource_ && !dirtyState_.getValue())
+                  {
+                     dirtyState_.markDirty(false);
+                     source_.setSourceDocumentDirty(
+                           docUpdateSentinel_.getId(), true, 
+                           new VoidServerRequestCallback());
+                  }
+               }  
+            });
+            
+            // save selection
+            panmirror_.addSelectionChangeHandler(new SelectionChangeEvent.Handler()
+            {
+               @Override
+               public void onSelectionChange(SelectionChangeEvent event)
+               {
+                  saveLocationOnIdle_.nudge();
+               }
+            });
+             
+            // track changes in outline sidebar and save as prefs
+            panmirror_.addPanmirrorOutlineVisibleHandler((event) -> {
+               setOutlineVisible(event.getVisible());
+            });
+            panmirror_.addPanmirrorOutlineWidthHandler((event) -> {
+               setOutlineWidth(event.getWidth());
+            });
+    
+            // good to go!
+            ready.execute();
+         });
+      }
+      else
+      {
+         // panmirror already created
+         ready.execute();
+      }
+   } 
+   
+   private String getEditorCode()
+   {
+      TextEditorContainer editorContainer = display_.editorContainer();
+      TextEditorContainer.Editor editor = editorContainer.getEditor();
+      return editor.getCode();
+   }
+   
+   // is our widget active in the editor container
+   private boolean isPanmirrorActive()
+   {
+      return display_.editorContainer().isWidgetActive(panmirror_);
+   }
+   
+   private TextEditorContainer.Editor getSourceEditor()
+   {
+      return display_.editorContainer().getEditor();
+   }
+  
+   private boolean getOutlineVisible()
+   {
+      return target_.getPreferredOutlineWidgetVisibility();
+   }
+   
+   private void setOutlineVisible(boolean visible)
+   {
+      target_.setPreferredOutlineWidgetVisibility(visible);
+   }
+   
+   private double getOutlineWidth()
+   {
+      return target_.getPreferredOutlineWidgetSize();
+   }
+   
+   private void setOutlineWidth(double width)
+   {
+      target_.setPreferredOutlineWidgetSize(width);
+   }
+   
+   
+   private PanmirrorEditingLocation savedEditingLocation()
+   {
+      String location = docUpdateSentinel_.getProperty(RMD_VISUAL_MODE_LOCATION, null);
+      if (StringUtil.isNullOrEmpty(location))
+         return null;
+      
+      String[] parts = location.split(":");
+      if (parts.length != 2)
+         return null;
+      
+      try
+      {
+         PanmirrorEditingLocation editingLocation = new PanmirrorEditingLocation();
+         editingLocation.pos = Integer.parseInt(parts[0]);
+         editingLocation.scrollTop = Integer.parseInt(parts[1]);
+         return editingLocation;
+      }
+      catch(Exception ex)
+      {
+         return null;
```

Comment:
Should we LogDebugException here?

Reply:
Yup. done here: https://github.com/rstudio/rstudio/commit/3f0ba5f25efb59373eddc649d15a882357dbf31a